### PR TITLE
kompute: add backend registry / device interfaces and Q4_K shader

### PR DESCRIPTION
**Pull request by slp**
Get in line with the other backends by supporting the newer backend/device registry interfaces. While there, also add mul_mat_q4_k shader.

- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [ ] Low
  - [x] Medium
  - [ ] High

### CODE REVIEW

File: ggml/src/ggml-kompute.cpp [hunk 1]
```diff
@@ -2006,7 +2033,7 @@ ggml_backend_t ggml_backend_kompute_init(int device) {
     ggml_backend_t kompute_backend = new ggml_backend {
         /* .guid      = */ ggml_backend_kompute_guid(),
         /* .interface = */ kompute_backend_i,
-        /* .device    = */ nullptr,
+        /* .device    = */ ggml_backend_reg_dev_get(ggml_backend_kompute_reg(), 0),
```
Comment (slaren) [inline ggml/src/ggml-kompute.cpp:2036]: It's important that the device is set to the correct one

File: ggml/src/ggml-kompute.cpp [hunk 2]
```diff
@@ -1921,7 +1958,7 @@ ggml_backend_buffer_type_t ggml_backend_kompute_buffer_type(int device) {
         for (const auto & dev : devices) {
             vec.push_back({
                 /* .iface   = */ ggml_backend_kompute_buffer_type_interface,
-                /* .device  = */ nullptr,
+                /* .device  = */ ggml_backend_reg_dev_get(ggml_backend_kompute_reg(), 0),
```
Comment (slaren) [inline ggml/src/ggml-kompute.cpp:1961]: Same issue here.

**Review by slaren — COMMENTED** [review]
I cannot review the other changes, I would suggest keeping them in a separate PR.

File: ggml/src/ggml-kompute.cpp (hunk 1 above)
Comment (slp) [reply ggml/src/ggml-kompute.cpp]: Fixed, thanks.

**Comment by slp** [comment]
> I cannot review the other changes, I would suggest keeping them in a separate PR.

Thanks for the review. I've fixed the issues, cleaned up the code a little bit more, and dropped the commit adding the additional shader to deal with it in another PR.

**Review by slaren — APPROVED** [review]
The implementation of the backend reg and device interfaces looks good, can be merged after resolving the conflicts from #10026.

**Comment by slp** [comment]
> The implementation of the backend reg and device interfaces looks good, can be merged after resolving the conflicts from #10026.

Thanks, I've just rebased it on top of current master.

**Pull request merged**